frontend/src/pages/ExperimentDetails.test.tsx

```diff
@@ -0,0 +1,453 @@
+import TestUtils from '../TestUtils';
+import { shallow } from 'enzyme';
+import { Apis } from '../lib/Apis';
+import { PageProps } from './Page';
+import { range } from 'lodash';
+import { ApiExperiment } from '../apis/experiment';
+import { ApiResourceType } from '../apis/job';
+import { RoutePage } from '../components/Router';
+import { ToolbarProps } from '../components/Toolbar';
+import { QUERY_PARAMS } from '../lib/URLParser';
+
+describe('ExperimentDetails', () => {
+  const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation(() => null);
+  const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => null);
+
+  const updateToolbarSpy = jest.fn();
+  const updateBannerSpy = jest.fn();
+  const updateDialogSpy = jest.fn();
+  const updateSnackbarSpy = jest.fn();
+  const historyPushSpy = jest.fn();
+  const getExperimentSpy = jest.spyOn(Apis.experimentServiceApi, 'getExperiment');
+  const listJobsSpy = jest.spyOn(Apis.jobServiceApi, 'listJobs');
+  const listRunsSpy = jest.spyOn(Apis.runServiceApi, 'listRuns');
+
+  const MOCK_EXPERIMENT = newMockExperiment();
+
+  function newMockExperiment(): ApiExperiment {
+    return {
+      description: 'mock experiment description',
+      id: 'some-mock-experiment-id',
+      name: 'some mock experiment name',
+    };
+  }
+
+  function generateProps(): PageProps {
+    return {
+      history: { push: historyPushSpy } as any,
+      location: '' as any,
+      // 'eid' here corresponds to RouteParams.experimentId 
+      match: { params: { eid: MOCK_EXPERIMENT.id } } as any,
+      toolbarProps: ExperimentDetails.prototype.getInitialToolbarState(),
+      updateBanner: updateBannerSpy,
+      updateDialog: updateDialogSpy,
+      updateSnackbar: updateSnackbarSpy,
+      updateToolbar: updateToolbarSpy,
+    };
+  }
+
+  async function mockNJobs(n: number): Promise<void> {
+    listJobsSpy.mockImplementation(() => ({
+      jobs: range(n).map(i => ({ id: 'test-job-id' + i, enabled: true, name: 'test job name' + i })),
+    }));
+    await listJobsSpy;
+    await TestUtils.flushPromises();
+  }
+
+  async function mockNRuns(n: number): Promise<void> {
+    listRunsSpy.mockImplementation(() => ({
+      runs: range(n).map(i => ({ id: 'test-run-id' + i, name: 'test run name' + i })),
+    }));
+    await listRunsSpy;
+    await TestUtils.flushPromises();
+  }
+
+  beforeEach(async () => {
+    // Reset mocks
+    consoleLogSpy.mockReset();
+    consoleErrorSpy.mockReset();
+    updateBannerSpy.mockReset();
+    updateDialogSpy.mockReset();
+    updateSnackbarSpy.mockReset();
+    updateToolbarSpy.mockReset();
+    getExperimentSpy.mockReset();
+    historyPushSpy.mockReset();
+    listJobsSpy.mockReset();
+    listRunsSpy.mockReset();
+
+    getExperimentSpy.mockImplementation(() => newMockExperiment());
+
+    await mockNJobs(0);
+    await mockNRuns(0);
+  });
+
+  it('renders a page with no runs or recurring runs', async () => {
```

Comment:
If you expect no errors here, can we assert that `updateBannerSpy` wasn't called (after initial call)?